src/openrct2/rct2/SeaDecrypt.cpp

```diff
@@ -0,0 +1,109 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2020 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "../common.h"
+#include "../core/File.h"
+#include "../core/Path.hpp"
+#include "RCT2.h"
+
+#include <cstdint>
+#include <memory>
+#include <string_view>
+
+constexpr int32_t MASK_SIZE = 0x1000;
+
+struct EncryptionKey
+{
+    uint32_t Seed0{};
+    uint32_t Seed1{};
+};
+
+static EncryptionKey GetEncryptionKey(const std::string_view& fileName)
+{
+    auto fileNameLen = static_cast<int32_t>(fileName.size());
+    uint32_t s0 = 0;
+    for (int i = fileNameLen - 1; i >= 0; i--)
+    {
+        s0 = (s0 + (s0 << 5)) ^ fileName[i];
+    }
+
+    uint32_t s1 = 0;
+    for (int i = 0; i < fileNameLen; i++)
+    {
+        s1 = (s1 + (s1 << 5)) ^ fileName[i];
+    }
+
+    return EncryptionKey{ s0, s1 };
+}
+
+static std::unique_ptr<uint8_t[]> CreateMask(const EncryptionKey& key)
+{
+    auto result = std::make_unique<uint8_t[]>(MASK_SIZE);
+    auto dst8 = result.get();
+    uint32_t seed0 = key.Seed0;
+    uint32_t seed1 = key.Seed1;
+    int32_t i = MASK_SIZE;
+    while (i > 0)
+    {
+        uint32_t s0 = seed0;
+        uint32_t s1 = seed1 ^ 0xF7654321;
+        seed0 = rol32(s1, 25) + s0;
+        seed1 = rol32(s0, 29);
+        *dst8++ = (s0 >> 3) & 0xFF;
+        if (i >= 2)
+        {
+            *dst8++ = (s0 >> 11) & 0xFF;
+            if (i >= 3)
+            {
+                *dst8++ = (s0 >> 19) & 0xFF;
+                if (i >= 4)
+                {
+                    *dst8++ = (seed1 >> 24) & 0xFF;
+                    i -= 4;
+                }
+            }
+        }
+    }
+    return result;
+}
+
+static void Decrypt(std::vector<uint8_t>& data, const EncryptionKey& key)
+{
+    auto mask = CreateMask(key);
+    const uint8_t* mask8 = (const uint8_t*)mask.get();
```

Comment:
This pointer is not required.

Reply:
Forgetting about the vector for a moment, how do you index into a `unique_ptr<uint8_t[]>`?

Reply:
Should be possible with just the normal [] operator https://en.cppreference.com/w/cpp/memory/unique_ptr/operator_at